tempesta_fw/tls.c

```diff
@@ -71,12 +271,227 @@ static TfwConnHooks tls_conn_hooks = {
+	mbedtls_ssl_config_free(&tfw_tls.cfg);
+}
+
+/*
+ * ------------------------------------------------------------------------
+ *	configuration handling
+ * ------------------------------------------------------------------------
+ */
+
+static int
+tfw_tls_cfg_start(void)
+{
+	int r;
+
+	if ((tfw_tls.crt.version && !tfw_tls.key.pk_ctx) ||
+	    (!tfw_tls.crt.version && tfw_tls.key.pk_ctx)) {
+		TFW_ERR("TLS: SSL certificate/key pair is incomplete\n");
+		return -EINVAL;
+	}
+
+	mbedtls_ssl_conf_ca_chain(&tfw_tls.cfg, tfw_tls.crt.next, NULL);
+	r = mbedtls_ssl_conf_own_cert(&tfw_tls.cfg, &tfw_tls.crt, &tfw_tls.key);
+	if (r) {
+		TFW_ERR("TLS: can't set own certificate (%x)\n", -r);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static void
+tfw_tls_cfg_stop(void)
+{
+	mbedtls_x509_crt_free(&tfw_tls.crt);
+	mbedtls_pk_free(&tfw_tls.key);
+}
+
+/**
+ * Handle 'ssl_certificate <path>' config entry.
+ */
+static int
+tfw_tls_cfg_handle_crt(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	int r;
+	void *crt_data;
+	size_t crt_size;
+
+	mbedtls_x509_crt_init(&tfw_tls.crt);
+
+	if (ce->attr_n) {
+		TFW_ERR("%s: Arguments may not have the \'=\' sign\n",
+			cs->name);
+		return -EINVAL;
+	}
+	if (ce->val_n != 1) {
+		TFW_ERR("%s: Invalid number of arguments: %d\n",
+			cs->name, (int)ce->val_n);
+		return -EINVAL;
+	}
+
+	crt_data = tfw_cfg_read_file((const char *)ce->vals[0], &crt_size);
+	if (!crt_data) {
+		TFW_ERR("%s: Can't read certificate file '%s'\n",
+			ce->name, (const char *)ce->vals[0]);
+		return -EINVAL;
+	}
+
+	r = mbedtls_x509_crt_parse(&tfw_tls.crt,
+				   (const unsigned char *)crt_data,
+				   crt_size);
+	vfree(crt_data);
+
+	if (r) {
+		TFW_ERR("%s: Invalid certificate specified (%x)\n",
+			cs->name, -r);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**
+ * Handle 'ssl_certificate_key <path>' config entry.
+ */
+static int
+tfw_tls_cfg_handle_crt_key(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	int r;
+	void *key_data;
+	size_t key_size;
+
+	mbedtls_pk_init(&tfw_tls.key);
+
+	if (ce->attr_n) {
+		TFW_ERR("%s: Arguments may not have the \'=\' sign\n",
+			cs->name);
+		return -EINVAL;
+	}
+	if (ce->val_n != 1) {
+		TFW_ERR("%s: Invalid number of arguments: %d\n",
+			cs->name, (int)ce->val_n);
+		return -EINVAL;
+	}
+
+	key_data = tfw_cfg_read_file((const char *)ce->vals[0], &key_size);
+	if (!key_data) {
+		TFW_ERR("%s: Can't read certificate file '%s'\n",
+			ce->name, (const char *)ce->vals[0]);
+		return -EINVAL;
+	}
+
+	r = mbedtls_pk_parse_key(&tfw_tls.key,
+				 (const unsigned char *)key_data,
+				 key_size, NULL, 0);
+	vfree(key_data);
+
+	if (r) {
+		TFW_ERR("%s: Invalid private key specified (%x)\n",
+			cs->name, -r);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static TfwCfgSpec tfw_tls_cfg_specs[] = {
+	{
+		"ssl_certificate", NULL,
+		tfw_tls_cfg_handle_crt,
+		.allow_none = true,
+		.allow_repeat = false,
+	},
+	{
+		"ssl_certificate_key", NULL,
+		tfw_tls_cfg_handle_crt_key,
+		.allow_none = true,
+		.allow_repeat = false,
+	},
+	{}
```

Comment:
Please add appropriate description to README.md whenever your create new configuration options.